velox/exec/tests/SpillerTest.cpp

```diff
@@ -22,183 +22,323 @@
+    EXPECT_EQ(numSpilledFiles, spilledFileSet.size());
+
+    for (auto spilledFile : spilledFileSet) {
+      auto readFile = fs_->openFileForRead(spilledFile);
+      EXPECT_NE(readFile.get(), nullptr);
+    }
+
+    Spiller::SpillRows unspilledPartitionRows = spiller_->finishSpill();
+    if (spillPct == 100) {
+      EXPECT_TRUE(unspilledPartitionRows.empty());
+      EXPECT_EQ(0, rowContainer_->numRows());
+    }
+
+    verifySpillData();
+
+    spiller_.reset();
+    // Verify the spilled file has been removed from file system after spiller
+    // destruction.
+    for (auto spilledFile : spilledFileSet) {
+      EXPECT_ANY_THROW(fs_->openFileForRead(spilledFile));
+    }
+  }
+
+  void setupSpillData(
+      std::shared_ptr<const RowType> rowType,
+      int32_t numKeys,
+      int32_t numRows,
+      std::function<void(RowVectorPtr)> customizeData = {},
+      std::vector<int> numRowsPerPartition = {}) {
+    if (!numRowsPerPartition.empty()) {
+      int32_t totCount = 0;
+      for (auto count : numRowsPerPartition) {
+        totCount += count;
+      }
+      VELOX_CHECK_EQ(totCount, numRows);
+    }
+
+    rowVector_ = BaseVector::create<RowVector>(rowType, numRows, pool_.get());
+
+    const SelectivityVector allRows(numRows);
+
+    const auto& childTypes = rowType->children();
+    std::vector<TypePtr> keys(childTypes.begin(), childTypes.begin() + numKeys);
+    std::vector<TypePtr> dependents;
+    if (numKeys < childTypes.size()) {
+      dependents.insert(
+          dependents.end(), childTypes.begin() + numKeys, childTypes.end());
+    }
+    // Make non-join build container so that spill runs are sorted. Note
+    // that a distinct or group by hash table can have dependents if
+    // some keys are known to be unique by themselves. Aggregation
+    // spilling will be tested separately.
+    rowContainer_ = makeRowContainer(keys, dependents, false);
+
+    // Setup temporary row to check spilling partition number.
+    char* testRow = rowContainer_->newRow();
+    std::vector<char*> testRows(1, testRow);
+    const auto testRowSet = folly::Range<char**>(testRows.data(), 1);
+    std::vector<uint64_t> hashes(1);
+
+    int numFilledRows = 0;
+    do {
+      RowVectorPtr batch = makeDataset(rowType, numRows, customizeData);
+      if (!numRowsPerPartition.empty()) {
+        SelectivityVector selectedRows(numRows);
+        for (int index = 0; index < numRows; ++index) {
+          for (int i = 0; i < keys.size(); ++i) {
+            DecodedVector decodedVector(*batch->childAt(i), allRows);
+            rowContainer_->store(decodedVector, index, testRow, i);
+            // Calculate hashes for this batch of spill candidates.
+            rowContainer_->hash(i, testRowSet, i > 0, hashes.data());
+          }
+          const int partitionNum =
+              hashBits_.partition(hashes[0], numPartitions_);
+          if (numRowsPerPartition[partitionNum] == 0) {
+            selectedRows.setValid(index, false);
```

Comment:
selectedRows.setValid() calls must be followed by selectedRows.updateBounds()

```
  /**
   * Set whether given index is selected. updateBounds() need to be called
   * explicitly after setValid() call, it can be called only once after multiple
   * setValid() calls in a row.
   */
  void setValid(vector_size_t idx, bool valid) {
```

Reply:
Thanks!